### ISSUE TITLE
☂️ Reinforcement: Add more basic components to the core framework

### ISSUE THREAD

**Issue by Hawmex** [flutter/flutter#97496]
This issue tracks proposals and development for net-new widgets and functionalities to be added to the core widgets library.

This effort is the other side of the coin to our reinforcement work, see the [refactoring side](https://github.com/flutter/flutter/issues/53059) as well. While the refactoring issue focuses on moving existing code, this issue is dedicated to identifying and building new, foundational, style-agnostic primitives that are missing from the core widgets library today. The goal is to fill gaps in the core and make it an even more powerful foundation for building any design system.

> [!NOTE]
> Some of these may be best informed by a UXR study. Before considering contributing one of these net-new widgets, reach out to the team on Discord to confirm.

### Propose a New Core Widget! 💡

What foundational widget or capability do you wish existed in Flutter's core? If you have an idea for a new, un-opinionated "headless" widget that would be broadly useful, this is the place to propose it. Please provide a clear use case and rationale for why it belongs in the core framework.

### _Last update Mar 4, 2026_

So far, we have investigated many of the existing design components to determine if any refactoring can be done *before* decoupling, moving existing APIs to the widgets layer to a common surface (which is tracked in [https://github.com/flutter/flutter/issues/53059](https://github.com/flutter/flutter/issues/53059)). In some cases, we have decided not to break widgets in the design libraries to try to accomplish this, which leaves the opportunity for new, more generic APIs to be introduced in the widgets layer later.

As we have been focused on the opportunity to refactor existing widgets (#53059) ahead of actually separating the libraries, expect this work of adding _new_ widgets to continue *after* the design libraries land in their new packages, which is what this issue is meant to track.

To check out all the issues and aspects of the decoupling effort, see the [github project](https://github.com/orgs/flutter/projects/220), and the root parent issue: https://github.com/flutter/flutter/issues/101479.

#### Status:

- What's landed or being worked on wrt raw widgets:  
  - **Menus**  
    - [RawMenuAnchor](https://main-api.flutter.dev/flutter/widgets/RawMenuAnchor-class.html)  
    - [RawMenuAnchorGroup](https://main-api.flutter.dev/flutter/widgets/RawMenuAnchorGroup-class.html)  
    - [RawMenuOverlayInfo](https://main-api.flutter.dev/flutter/widgets/RawMenuOverlayInfo-class.html)  
  - **Radios**  
    - [RawRadio](https://main-api.flutter.dev/flutter/widgets/RawRadio-class.html)  
    - [https://github.com/flutter/flutter/pull/168161](https://github.com/flutter/flutter/pull/168161)  
    -   
  - **Tooltips**  
    - [RawTooltip](https://main-api.flutter.dev/flutter/widgets/RawTooltip-class.html)  
    - [https://flutter.dev/go/codeshare-tooltip](https://flutter.dev/go/codeshare-tooltip)   
  - **Page transitions**  
    - [https://github.com/flutter/flutter/issues/172929](https://github.com/flutter/flutter/issues/172929)   
    - [flutter.dev/go/decouple-page-transition-builders](http://flutter.dev/go/decouple-page-transition-builders)  
  - **Other**  
    - [Expansible](https://main-api.flutter.dev/flutter/widgets/Expansible-class.html)  
      - [https://flutter.dev/go/codeshare-expansion-tile](https://flutter.dev/go/codeshare-expansion-tile)   
    - showRawDialog  
      - [https://github.com/flutter/flutter/pull/181861](https://github.com/flutter/flutter/pull/181861) 
- What's been looked into and will revisit after decoupling lands:  
  - **TabBar** 
    - https://github.com/flutter/flutter/issues/175592 
    - Navigation oriented, or still scrolling PageView  
    - We found this API has a lot of material baked-in. Rather than breaking it, after decoupling lands, we'll revisit a generic TabBar in the widgets layer.  
  - **Theming**  
    - https://github.com/flutter/flutter/issues/177932
    - [design doc](https://docs.google.com/document/d/1hv3fjt-NYgiPRjBXF0Fb9TvFuaUtWbGhSMRX7-nrNnE/edit?usp=sharing), we'll revisit this after decoupling lands.  
    - We discussed a new base theme object as making sense. For use in the WidgetsApp, containing:  
      - ThemeExtensions, or something similar  
      - High contrast (a11y)  
      - Brightness (dark mode)  
      - defaultTargetPlatform  
      - Font family and fallback  
      - Perhaps others…  
  - **Text input and selection stack**  
    - [flutter.dev/go/decoupling-design-from-text](https://flutter.dev/go/decoupling-design-from-text)  
    - [https://github.com/flutter/flutter/issues/172930](https://github.com/flutter/flutter/issues/172930)   
    - We discussed:  
      - The future of SelectableText  
        - Here we discussed the state of text selection components in the widgets library without SelectableText after decoupling Material and Cupertino from the framework.  
        - The recommendation here was to move forward with SelectableText leaving the framework with Material, and to continue working on fixing critical bugs in SelectableRegion \+ Text.  
      - Design agnostic text selection UI  
        - This discussion was around whether it was feasible to move text selection UI such as context menus and selection handles to the widgets library.  
        - The recommendation here was to not move forward with text selection UI in the widgets library at this time due to the heavy dependencies on Material and Cupertino. Bringing these components into the widgets library would result in a less stable framework as it will still be susceptible to design changes.  
      - The default out of the box experience for text  
        - Here we discussed the API usability between the framework’s text and text input offerings with Material and Cupertino’s. Specifically we looked at EditableText vs TextField and SelectionArea vs SelectableRegion.  
        - We are not making any changes at the moment but have identified some areas to explore after decoupling.  
  - **Colors**  
    - [https://github.com/flutter/flutter/issues/172931](https://github.com/flutter/flutter/issues/172931)  
    - [flutter.dev/go/basic-color-set](http://flutter.dev/go/basic-color-set)   
- What's also up for consideration (want to add to the list? Add a comment with a suggestion\!):  
  - **Basic Button (wrapper)**  
    - Functionality, semantics over style, for use in composing anything to be a button, would contain:  
      - Focus  
      - Semantics  
      - Gestures  
      - MouseRegion  
    - Related: TestButton  
      - [https://github.com/flutter/flutter/blob/master/packages/flutter/test/widgets/button\_tester.dart](https://github.com/flutter/flutter/blob/master/packages/flutter/test/widgets/button_tester.dart)  
  - **Bottom bar**  
    - Similar to BottomAppBar/BottomNavigationBar, simpler generic interface, safe area handling  
  - **Top bar**  
    - Similar to App Bar, simpler generic interface, safe area handling

#### Raw- widgets that pre-date this effort

This is not an exhaustive list of components available in the [widgets](https://main-api.flutter.dev/flutter/widgets/) library, but specifically those with Raw prefixes. 🙂

- [RawAutocomplete](https://api.flutter.dev/flutter/widgets/RawAutocomplete-class.html)  
- [RawDialogRoute](https://api.flutter.dev/flutter/widgets/RawDialogRoute-class.html)  
- [RawGestureDetector](https://api.flutter.dev/flutter/widgets/RawGestureDetector-class.html)  
- [RawImage](https://api.flutter.dev/flutter/widgets/RawImage-class.html)  
- [RawMagnifier](https://api.flutter.dev/flutter/widgets/RawMagnifier-class.html)  
- [RawScrollbar](https://main-api.flutter.dev/flutter/widgets/RawScrollbar-class.html)  
- [RawView](https://main-api.flutter.dev/flutter/widgets/RawView-class.html)

---

<details> <summary> OP </summary> 
Hi,
It's been 2 months since I've ditched HTML, CSS, and JS for Flutter and I haven't looked back. I love the declarative style of Flutter and Dart itself.

### What's the problem?
What really bothers me is I find it hard to implement completely Custom UIs. I mean even when I wanted to use a custom font, I had to depend on `material.dart` and use `Scaffold` to make it work. I don't even want to start talking about the frustrations of theming a material app in flutter. But this was not a problem using HTML, CSS, JS. They basically give you simple and minimal elements and then let you do whatever you want with them or use whatever UI package you want, such as Tailwind, MUI, etc.
So, It's either a matter of Flutter lacking documentation on Custom UIs or me not reading enough docs about them.

### Any other problems?
Well, if you just take a look at the reported issues, they're mostly about `material.dart` and `cupertino.dart`. I feel like Flutter Team is spending most of their time, effort and energy on these two UIs instead of the framework itself. And still, Flutter is far behind from Material 3 spec (and even Material 2, because `material.dart` uses 2014 typography by default.).

### My Ideal Flutter
So my ideal is:
- The main Flutter package should contain and export basic widgets of the framework (`GestureDetector`, `StatelessWidget`, `StatefulWidget`, etc.). These can be currently found in `widgets.dart`. This way, Flutter Team's efforts are concentrated on the framework itself.
- `material` and `cupertino` should come in seperate packages (and export the main Flutter package) and be maintained by Flutter community more.
- More tutorials and docs should be provided on making UIs with the main Flutter package rather than making them with `material` or `cupertino`. This way more people will know how underlying widgets of `material` and `cupertino` work and can take part in maintaining them or even make their own UIs in their apps and packages.

This way we will be able to see more Flutter UI packages and a more diverse set of Flutter apps in terms of UI on the market.

After all, these are just my ideals and may be far from reality but I really like to see some discussion on this matter and finding the best solution to it.

</details>

**Comment by HerrNiklasRaab** [comment]
I agree with this so much!

**Comment by darshankawar** [comment]
@Hawmex 
Thanks for the report. Implementing custom UIs is probably a broad topic. Can you give us a real world example where custom UI you proposed would be helpful rather than using the current implementation ?

**Comment by Hawmex** [comment]
Well, decoupling `material.dart` from Flutter has been also discussed in #91605 and many reasons have somehow been ignored.

> FWIW, the main reason to not have material be its own package is that we want the out-of-box experience to be quick. That is, people should be able to download Flutter, and immediately start writing an app with pretty buttons.

> Yeah, but Flutter kind of already does that for icons and it works, right? There is `uses-material-design: true` and `cupertino_icons` in every app (because of the template) and you don't see people complaining it is hard to add material icons to Flutter because most use the template.

But the reason for implementing custom UIs is in my opinion an obvious thing. Not all developers and companies want to follow Cupertino or Material design guidelines. And even in their UI architecture, they may want to follow different architecture patterns from `material.dart` (for example theming, scaffolding, etc.).

I live in Iran and I'd like to use Persian fonts rather than the default Roboto. I followed the documentation but it didn't work. Turned out I had to wrap my page in `Scaffold`, but it wasn't mentioned in the docs.

Some other reasons:

- Even in `material.dart` itself, there are inconsistencies in coding patterns. Opening dialogs is handled through a top-level function while showing a snackbar is handled through an `InheritedWidget` pattern.
- Flutter has always had to avoid breaking changes in `material.dart` and go on with accumulating old stuff and deprecating them for God knows how long. Just take a look at how many parameters `ThemeData` has as of today.
- `material.dart` has been integrated with the framework itself that you can't make Flutter apps without it. A fine example would be the urge to use `Scaffold` to load custom fonts in the current theming pattern.

Just decouple `material.dart` and `cupertino.dart` from Flutter and give people the option to make their own UI or choose it from external packages including Material and Cupertino. And I'm sure those who love material design would continue to use it in a separate package.

**Comment by darshankawar** [comment]
Keeping it open and labeling as a proposal for further inputs from the team.

/cc @HansMuller

**Comment by darshankawar** [comment]
I just found out this similar / related issue: https://github.com/flutter/flutter/issues/79751
@Hawmex Please confirm if your proposal can be tracked in the linked issue.

**Comment by Hawmex** [comment]
@darshankawar
About *Providing documentation on the undrlying Flutter framework widgets*, yes. It can be tracked there.

But I also propose getting `material.dart` and `cupertino.dart` out of Flutter and providing them as separate packages.

**Comment by n7trd** [comment]
I absolutely agree on this one! Flutter is an awesome framework to build cross-platform UIs. 
Also like the current focus on expanding the platform support to desktop & web.

My only concern is the extensive focus on material-ui. While it is clear to me that its the preferred ui for mobile devices, i'm not so sure its the best choice for desktop, web and embedded UIs. If we look to the most recent apps like Rive, Rows, fluent_ui (lib) and the interest of companies in the embedded space, we see custom UIs everywhere. Even on mobile, flutter is used more for custom UIs that diverge from the material design system. I think there is a huge potential there. But there could be even more developers if, like @Hawmex stated, we would have

- more docs on implementing custom UIs
- more focus on the "core" ui stack
- seperated packages for material-ui and cupertino. (which would be even beneficial to material-ui #91605 & cupertino) and a clear statement for the diversity the framework is capable of.

**Comment by n7trd** [comment]
While the proposal to put material-ui and cupertino in seperated packages is clear, the other points are a bit vague. I want to give some examples how the current situation for custom UIs can be improved:

### Docs

- **The introduction to building UIs, like the basic introduction to widgets and layouts, shouldn't have references to material-ui and its widgets**. That removes lots of "in an material app, you do this and that". That makes the documentation more clear and keeps the focus on the basics. Why shoud i care about ElevatedButton or FloatingActionButton in the first tutorials about widgets and layout?! 

- **Add separated tutorials/guides for material-ui, cupertino and references to other community UI packages.** This makes it also clear that material-ui is only _one_ way to build UIs with flutter. A good example is the documentation about state management. Basic introduction to core concepts -> Overview about available options.

- **Add guides/best practices/tutorials how to implement custom widgets.** I think there is lots of knowledge from the flutter team the developers can benefit from. There is also a lack of information about more advanced topics (paint/canvas/...). In case someone wants to create widgets that can not be composed of existing core widgets.

### More focus on the "core UI"

To be honest, there is already lots of functionality in libraries like widgets, painting, gestures,... Maybe its just the lack of guides to find an entry.

But it would definitely be beneficial to get feedback from the community about whats missing here. Some areas could be: **Primitives, helpers or more higher level components for code reuse across UI libraries to reduce boilerplate code.**

**Comment by n7trd** [comment]
It also looks like the api-reference, especially the way things get re-exported in libraries, makes 'learning from the api' 
unnecessary difficult. I created a separated issue #100444 for that.

**Comment by HansMuller** [comment]
CC @leighajarett

**Comment by AliYar-Khan** [comment]
I agree with making the material and Cupertino a choice for developers not enforced by the framework/SDK!!!

**Comment by NatanAmorim** [comment]
I'm really trying to post things were they are relevant, I'm sorry if something I posted here would be more relevant somewhere else.

I think moving the Material and Cupertino packages outside of Flutter, like it was proposed in #101479, would be great, I also made some comments about the topic in <https://github.com/flutter/flutter/issues/165502#issuecomment-2966731394>, Flutter could focus on a better Standard for Implementing Custom UIs. I love what Flutter has done so far, and the code quality of what we have now is astonishing! but i'm afraid of what's going to happen in the future.

Flutter started with the focus of building Android and IOS Apps, Flutter proved itself as a great tool for cross-platform mobile app development. But now Flutter has expanded, it's on Desktop, it's on Web and even on embedded devices, Flutter is getting closer to 10 years of success. But even with the community help, changes need to be approved and due to the high standard of code quality, the changes and approval takes time.

Gluing Material and Cupertino with flutter has benefits and problems, a recent problem was:

> Finally, Flutter 3.27 has a number of rendering bugs and crashes related to Impeller rendering on Vulkan capable devices that are fixed in 3.29 and later versions. We do not plan to hotfix any of these changes into 3.27 and we highly recommend updating to 3.29 or later.

Source: [What’s new in Flutter 3.32 (Medium)](https://medium.com/flutter/whats-new-in-flutter-3-32-40c1086bab6e)

You need a update to solve this problem, but if something else in Material and Cupertino changed (in this update or other updates) it can cause situations where teams take longer to get updates that fix bugs and crashes, specially if the team has a slow process it needs to follow to update their apps. If you move the Material and Cupertino to packages outside of the main Flutter code you can update and fix bugs or security issues with the benefit of dealing with the UI whenever you want.

Recently, in Google I/O 2025 Flutter showed it's work on [squircles](https://youtu.be/v6Rzo5khNE8?t=1366) in it's efforts to faithfully replicate Apple's aesthetic with highest possible fidelity, this was a couple of weeks before Apple announced Liquid Glass (Youtube: [Introducing Liquid Glass | Apple](https://youtu.be/jGztGfRujSE)), so far I've only used Material Design I have no Idea how much of the Flutter team incredible work on Cupertino is now obsolete after years of immense dedication, I cannot imagine how long it would take to make a Liquid Glass implementation feel like native on Apple devices, one example of how complex it is to implement (from #170310):

> I think the main problem with implementing Liquid Glass in Flutter is not handling sampling the background of a widget - because as you've shown, that's absolutely doable, even if at the cost of some performance - but rather having the interactions between separate widgets (for example stuff like two separate buttons blending together during a transition). I imagine you'd need to rethink how Flutter renders widgets, and instead of having each widget have its own render object, you'd need a single layer somewhere higher up the tree, that all the liquid glass widgets would report their info to. That layer would then need to compile all that into a single fragment shader, that would render all those effects using SDFs and apply those refraction and specular lighting effects.

Source: <https://github.com/flutter/flutter/issues/170310#issuecomment-2965431804>

I think Flutter should focus on making things possible, but the implementation of design languages should be the responsibility of the community. But, it would be good to still have an official implementation of something like Material Design to help dogfooding "A Standard For Implementing Custom UIs" and because the Flutter team is part of Google (at least I think that could help implementing MD). We could have Material Design as a official UI implementation of a new standard, being developed by the Flutter team, it would be a good UI for those that just want to use an UI no matter which one since Material Desing also is very well documented, but would also be easier for those that want to use something else, while also allowing other to create their on design languages without a lot of headaches.

Also here's another comment from #170310 that I liked and think are more relevant here:

> First, let's be clear about what Flutter's core responsibility is.
As you said, the Flutter team's primary task is to guarantee the stability, performance, and cross-platform consistency of its rendering engine (Skia/Impeller). Flutter's core promise is that a single codebase you write can be rendered precisely and consistently on any platform it supports—be it iOS, Android, Windows, or the Web—while delivering performance that is as smooth as native.
> These issues are the bedrock of the framework. A rendering error, a single instance of jank, or a compilation problem has a far wider and more severe impact than the absence of a specific UI effect. Faced with these fundamental problems that affect all developers and the stability of all apps, allocating precious development resources to implement a specific, even "fanciful" UI effect, is clearly not a wise or responsible decision. The Flutter team must prioritize ensuring the foundation is solid before getting busy decorating the rooftop.
> Flutter's success is half due to its powerful core engine and half due to its vibrant open-source community and the pub.dev ecosystem. The Flutter team is responsible for building the solid "engine" and "chassis," while the community is empowered to create all sorts of beautiful "car bodies" and "interiors."
>
> This isn't an abdication of responsibility; it's a healthy and efficient ecosystem model. If you are passionate about the Liquid Glass effect, the best path forward is to create a high-quality package and publish it on pub.dev. Not only would this immediately solve your own need, but it would also be a tremendous contribution to the entire community. A well-designed, performant Glass component package, once published, would be welcomed by all of us, and we would be grateful for your contribution. This is a much faster path than waiting for the official team to even consider this feature after they've addressed thousands of core issues.

Source: <https://github.com/flutter/flutter/issues/170310#issuecomment-2964952689>

##### Venting a little (feel free to ignore this)

This was a lot of text and I must admit part of what bothers me and made me want to write this, was because a feature that _to me_ **feels essential** to have in Desktop and Web UI, is [2D Scrolling](https://youtu.be/UDZ0LPQq-n8) with **"support for corner space with dual scrollbar"** #122345, I have been writing desktop apps with Flutter since 2022 (mobile since 2020) and even posted some very _ugly code_ ([(stackoverflow) Flutter web - Always show horizontal and vertical scrollbar in a DataTable](https://stackoverflow.com/questions/68899502/flutter-web-always-show-horizontal-and-vertical-scrollbar-in-a-datatable/70689232#70689232)) to have dual scrollbar even before the development of an official solution was announced, now years later [two_dimensional_scrollables](https://pub.dev/packages/two_dimensional_scrollables) is great, but, since the Flutter team has a lot to do, resources are limited and not a lot of people are asking for this, this will still take a long time to be done. I'm not asking things to be done here, I'm no one, but certainly feels frustrating when you are waiting something for years and see a huge undertaking to add something that looks to me unnecessary (Squircles) when something that feels essential is unfinished for years... (once again, unfinished but great, no, actually amazing work so far! 😃)

I would like to thank everyone who contributes to Flutter and say I love every day I can make something in Flutter because the framework and it's tools are perfect for me.

**Comment by Piinks** [comment]
Hi folks! We have decided to pursue decoupling the design libraries from the framework. Please see OP above for how we are coordinating work to improve the offering of core widgets.
Parent tracking issue: https://github.com/flutter/flutter/issues/101479

**Comment by s0nerik** [comment]
It looks like the list of components to consider lacks editable text field machinery. I'm talking about higher-level, design-agnostic components for text selection handling and related popup menus.

**Comment by Piinks** [comment]
> It looks like the list of components to consider lacks editable text field machinery. I'm talking about higher-level, design-agnostic components for text selection handling and related popup menus.

Good call out @s0nerik. As mentioned above, there are two sides to reinforcing the core framework:
1. refactoring logic out of existing components in the design libraries
2. adding missing components.

This issue is the later. See https://github.com/flutter/flutter/issues/53059 for the former, which includes text editing controls and selection. :)
More specifically - https://github.com/flutter/flutter/issues/172930

**Comment by davidhicks980** [comment]
Here are some widgets that I've used in multiple previous apps. Some of these are probably too specific to their use case, but hopefully something can be drawn from them:

<details><summary>Widgets</summary>

### Inert

Sometimes, it's nice to prevent widget rebuilds in response to media query changes. The example I always think of is underlying pages that listen to MediaQuery.paddingOf(context), but are not currently visible. I use the following `Inert` widget to prevent unnecessary rebuilds

<details><summary>Inert Widget</summary>

```dart
/// A widget that hides its child and prevents rebuilds in response to MediaQuery changes.
class Inert extends StatefulWidget {
  const Inert({super.key, required this.inert, required this.child});
  final bool inert;
  final Widget child;

  @override
  State<Inert> createState() => _InertState();
}

class _InertState extends State<Inert> {
  MediaQueryData? _mediaQuery;

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
    _mediaQuery = widget.inert && _mediaQuery != null ? _mediaQuery : MediaQuery.of(context);
  }

  @override
  void didUpdateWidget(Inert oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (widget.inert != oldWidget.inert) {
      _mediaQuery = widget.inert && _mediaQuery != null ? _mediaQuery : MediaQuery.of(context);
    }
  }

  @override
  Widget build(BuildContext context) {
    return Visibility(
      visible: !widget.inert,
      maintainState: true,
      child: MediaQuery(
        data: _mediaQuery!,
        child: widget.child,
      ),
    );
  }
}

```
</details>

### AnimatedDecoratedBox

Currently, DecoratedBoxTransition exists but AnimatedDecoratedBox does not. I implement it in every project, so it'd be nice to have from the core library. 

### Horizontal ListWheelScrollView

ListWheelScrollView currently only works in the vertical direction, enabling CupertinoPicker behavior:

https://github.com/user-attachments/assets/9f72fc83-aa10-4e5d-84e5-82e37dcae747

I was working on a podcast app for myself (RIP Google Podcasts), and came across a few situations where it would be helpful to have horizontal scrolling. For example, this Google Podcasts-inspired speed selector required me to fork ListWheelScrollView to add horizontal behavior. ListWheelScrollView is currently hardcoded to behave vertically:

https://github.com/user-attachments/assets/9e072456-b37b-42a7-9018-39f5d19c40c5

### Range slider (_Vertical and Horizontal_) 
Self explanatory.  I also think it'd be good to allow the knob to be a separate widget.

### ThresholdChild

Sometimes, you only want to show a widget if another widget's contents exceed a specified height. For example, in the following video, the paragraph is only expandable if it exceeds 100 px:

https://github.com/user-attachments/assets/acf3e93e-819d-47ab-a453-6f87e5a60d0d

Podcast app usage:

https://github.com/user-attachments/assets/d4b8b195-4f83-4f04-a79d-0f3208970c4d

### Theming

Curious if anyone has a better method of achieving Raw widget theming. I've been using a WidgetStatesController combined with an inherited notifier (WidgetStatesScope) and inherited widgets (WidgetsStatesTextStyle and WidgetStatesDecoration) to separate theming from core widgets. This is hard to explain, so here's an example using a raw button:

https://github.com/user-attachments/assets/56fb2b34-3bc2-467e-94bb-d7e167018102

<details><summary>RawButton theming demo</summary>

```dart
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const RawButtonApp());
}

class RawButtonApp extends StatelessWidget {
  const RawButtonApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Threshold Demo',
      theme: ThemeData(primarySwatch: Colors.blue),
      home: const Scaffold(body: Center(child: RawButtonDemo())),
    );
  }
}

class RawButtonDemo extends StatefulWidget {
  const RawButtonDemo({super.key});

  @override
  State<RawButtonDemo> createState() => _RawButtonDemoState();
}

class _RawButtonDemoState extends State<RawButtonDemo> with TickerProviderStateMixin {
  static final decoration = WidgetStateProperty.fromMap({
    WidgetState.pressed: const BoxDecoration(color: Colors.blue),
    WidgetState.hovered: const BoxDecoration(color: Colors.indigo),
    WidgetState.any: BoxDecoration(color: Colors.indigo[100]),
  });

  static final textStyle = WidgetStateProperty.fromMap({
    WidgetState.hovered: const TextStyle(fontWeight: FontWeight.w600, color: Colors.white),
    WidgetState.any: TextStyle(color: Colors.indigo[900], fontWeight: FontWeight.w600),
  });

  @override
  Widget build(BuildContext context) {
    return Column(
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        RawButton(
          onPressed: () {
            ScaffoldMessenger.of(context).showSnackBar(
              const SnackBar(content: Text('Button Pressed!')),
            );
          },
          // Note that decoration and text style are not provided to the button itself. WidgetStatesDecoration and WidgetStatesDefaultTextStyle respond to changes from the nearest inherited WidgetStatesScope
          child: WidgetStatesDecoration(
            decoration: decoration,
            child: WidgetStatesDefaultTextStyle(
              textStyle: textStyle,
              child: const Padding(
                padding: EdgeInsets.all(16.0),
                child: Text('Press Me'),
              ),
            ),
          ),
        ),
      ],
    );
  }
}
```
</details>

<details><summary>Theming implementation</summary>

```dart

// Widget states are passed using an inherited widget to descendent WidgetStatesDecoration and WidgetStatesDefaultTextStyles
class WidgetStateScope extends InheritedNotifier<WidgetStatesController> {
  const WidgetStateScope({
    super.key,
    required super.notifier,
    required super.child,
  });

  @override
  WidgetStatesController? get notifier => super.notifier;

  static Set<WidgetState>? maybeOf(BuildContext context) {
    return context.dependOnInheritedWidgetOfExactType<WidgetStateScope>()?.notifier?.value;
  }
}

class WidgetStatesDecoration extends StatelessWidget {
  const WidgetStatesDecoration({
    super.key,
    required this.child,
    required this.decoration,
  });

  final Widget child;
  final WidgetStateProperty<Decoration> decoration;

  @override
  Widget build(BuildContext context) {
    final state = WidgetStateScope.maybeOf(context)!;
    final resolvedDecoration = decoration.resolve(state);
    return DecoratedBox(
      decoration: resolvedDecoration,
      child: child,
    );
  }
}

class WidgetStatesDefaultTextStyle extends StatelessWidget {
  const WidgetStatesDefaultTextStyle({
    super.key,
    required this.child,
    required this.textStyle,
  });

  final Widget child;
  final WidgetStateProperty<TextStyle> textStyle;

  @override
  Widget build(BuildContext context) {
    final state = WidgetStateScope.maybeOf(context)!;
    final resolvedTextStyle = textStyle.resolve(state);
    return DefaultTextStyle(
      style: resolvedTextStyle,
      child: child,
    );
  }
}
```
</details>


<details><summary>RawButton implementation</summary>

```dart

class RawButton extends StatefulWidget {
  const RawButton({
    super.key,
    required this.child,
    this.onPressed,
    this.focusNode,
    this.autofocus = false,
    this.onFocusChange,
    String? focusNodeLabel,
    this.controller,
    this.target,
    this.alignment,
    this.behavior = HitTestBehavior.deferToChild,
  }) : _focusNodeLabel = focusNodeLabel;

  final Widget child;
  final VoidCallback? onPressed;
  final void Function(bool)? onFocusChange;
  final FocusNode? focusNode;
  final bool autofocus;
  final String? _focusNodeLabel;
  final WidgetStatesController? controller;
  final BoxConstraints? target;
  final AlignmentGeometry? alignment;
  final HitTestBehavior behavior;

  @override
  State<RawButton> createState() => _RawButtonState();
}

class _RawButtonState extends State<RawButton> {
  late final Map<Type, Action<Intent>> _actions = <Type, Action<Intent>>{
    ActivateIntent: CallbackAction<ActivateIntent>(onInvoke: _activateOnIntent),
    ButtonActivateIntent: CallbackAction<ButtonActivateIntent>(onInvoke: _activateOnIntent),
  };
  FocusNode get _focusNode => widget.focusNode ?? _internalFocusNode!;
  FocusNode? _internalFocusNode;
  WidgetStatesController? _internalController;
  WidgetStatesController get _controller => widget.controller ?? _internalController!;
  Set<WidgetState>? _states;

  late final gestures = <Type, GestureRecognizerFactory>{
    TapGestureRecognizer: GestureRecognizerFactoryWithHandlers<TapGestureRecognizer>(
      () => TapGestureRecognizer(),
      (TapGestureRecognizer instance) {
        instance
          ..onTapDown = _handleTapDown
          ..onTapUp = _handleTapUp
          ..onTapCancel = _handleTapCancel;
      },
    ),
  };

  @override
  void initState() {
    super.initState();
    if (widget.focusNode == null) {
      _internalFocusNode = FocusNode(debugLabel: widget._focusNodeLabel);
    }
    if (widget.controller == null) {
      _internalController = WidgetStatesController(_states);
    }
    _controller.addListener(_handleStateChange);
  }

  @override
  void didUpdateWidget(RawButton oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (oldWidget.focusNode != widget.focusNode) {
      if (widget.focusNode == null) {
        _internalFocusNode = FocusNode(debugLabel: widget._focusNodeLabel);
      } else {
        _internalFocusNode?.dispose();
        _internalFocusNode = null;
      }
    }

    if (oldWidget.controller != widget.controller) {
      _controller.removeListener(_handleStateChange);
      if (widget.controller == null) {
        _internalController = WidgetStatesController(_states);
      } else {
        _internalController?.dispose();
        _internalController = null;
      }
      _controller.addListener(_handleStateChange);
    }
  }

  @override
  void dispose() {
    _controller.removeListener(_handleStateChange);
    _internalController?.dispose();
    _internalController = null;
    _internalFocusNode?.dispose();
    super.dispose();
  }

  void _handleStateChange() {
    _states = _controller.value;
  }

  void _activateOnIntent(Intent intent) {
    _handlePressed();
  }

  void _handlePressed() {
    widget.onPressed?.call();
  }

  void _handleTapDown(TapDownDetails details) {
    _controller.update(WidgetState.pressed, true);
  }

  void _handleFocusChange(bool value) {
    _controller.update(WidgetState.focused, value);
    widget.onFocusChange?.call(value);
  }

  void _handleExit(PointerExitEvent event) {
    _controller
      ..update(WidgetState.hovered, false)
      ..update(WidgetState.pressed, false);
  }

  void _handleHover(PointerHoverEvent event) {
    _controller.update(WidgetState.hovered, true);
  }

  void _handleTapUp(TapUpDetails details) {
    _controller.update(WidgetState.pressed, false);
    _handlePressed.call();
  }

  void _handleTapCancel() {
    _controller.update(WidgetState.pressed, false);
  }

  @override
  Widget build(BuildContext context) {
    final Widget child = WidgetStateScope(notifier: _controller, child: widget.child);
    return Actions(
      actions: _actions,
      child: Focus(
        debugLabel: widget._focusNodeLabel,
        onFocusChange: _handleFocusChange,
        autofocus: widget.autofocus,
        focusNode: _focusNode,
        child: MouseRegion(
          onHover: _handleHover,
          onExit: _handleExit,
          hitTestBehavior: HitTestBehavior.deferToChild,
          child: RawGestureDetector(
            gestures: gestures,
            behavior: widget.behavior,
            child: widget.target != null
                ? ConstrainedBox(
                    constraints: widget.target!,
                    child: ColoredBox(
                      color: const Color(0x00000000), // Ensure the box is visible for hit-testing.
                      child: Align(
                        alignment: widget.alignment ?? Alignment.centerLeft,
                        child: child,
                      ),
                    ),
                  )
                : child,
          ),
        ),
      ),
    );
  }
}
```
</details>
</details>

**Comment by victorsanni** [comment]
> ## ThresholdChild
> Sometimes, you only want to show a widget if another widget's contents exceed a specified height. For example, in the > following video, the paragraph is only expandable if it exceeds 100 px:

Hi @davidhicks980, does the new [Expansible](https://api.flutter.dev/flutter/widgets/Expansible-class.html) widget solve this use case?

**Comment by davidhicks980** [comment]
@victorsanni No, because Expansible doesn't change its rendering based on the size of its body. You'd have to use a multi child render object that measures its body, checks if the body is larger than a given threshold (e.g. 100 px), and then renders the threshold child if that threshold is met. An example from the wild (quora.com):

<img width="400" alt="Image" src="https://github.com/user-attachments/assets/5ee70219-1d52-41d3-8649-a2704d5d7ad6" />

Lisa's comment renders a "Continue reading" because her comment exceeds a threshold height. Her comment is also constrained to that threshold height.

<img width="400" alt="Image" src="https://github.com/user-attachments/assets/c342835a-7c9d-4803-89e1-b77a14376bf9" />

Mary's comment doesn't exceed the threshold, so a "Continue reading" button is not shown. Her comment is not vertically constrained since it does not exceed the threshold height.

I'm pretty sure this isn't possible without a render object, but it may be possible using a Stack and a Layout Builder? I could be wrong.

Here's a rough example using a render object: https://dartpad.dev/ba613e9606a391c3da51207cea31e932 . I'm not sure if this is too specific of a use case, but it's one of the few times I've had to write a custom render object, so I thought it may be relevant.

**Comment by Piinks** [comment]
Hi everyone!
We've been so heads down on this work, if you haven't been keeping up with the individual issues, pull requests, and design documents, I thought an update was needed here. :)

This issue represents one of the 4 pillars of this effort (see [parent issue](https://github.com/flutter/flutter/issues/101479)), adding new basic components to the core framework.

During this period that the libraries are still together, we've been primarily focused on refactoring existing design widgets to those basic components where we can (tracked in https://github.com/flutter/flutter/issues/53059), but as a result, we already have some new components in the widgets layer! I have updated the OP to cover a lot of the work that has been done in relation to this pillar.

Once we land the libraries in their new packages (hello, material_ui and cupertino_ui 👋 ), we'll continue to work on this pillar of the project to add more to the core. Thanks all, keep the feedback coming!

**Comment by davidhicks980** [comment]
@Hannnes1 I've had luck using an InheritedNotifier that passes down widget state. 

For example, here's a Button that only defines interactions:

Dartpad: https://dartpad.dev/45ab05be46c7a72a2ee59d52fba1d6a8

(also, I just wrote Button as an example, so make sure to test your own version if you use this approach)

```dart

class ButtonExample extends StatelessWidget {
  const ButtonExample({super.key});

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Button(
        onPressed: () {
          ScaffoldMessenger.of(
            context,
          ).showSnackBar(const SnackBar(content: Text('Button Pressed!')));
        },
        behavior: HitTestBehavior.opaque,
        mouseCursor: WidgetStateProperty.all(SystemMouseCursors.click),
        child: Builder(
          builder: (context) {
            final ButtonState state = Button.of(context)!;
            final bool isHovered = state.hovered;
            final bool isPressed = state.pressed;
            final bool isFocused = state.focused;
            return AnimatedContainer(
              duration: const Duration(milliseconds: 50),
              padding: const EdgeInsets.symmetric(horizontal: 24, vertical: 12),
              decoration: BoxDecoration(
                color: isPressed
                    ? Colors.blue.shade700
                    : isHovered
                    ? Colors.blue.shade600
                    : Colors.blue.shade500,
                borderRadius: BorderRadius.circular(8),
                border: isFocused ? Border.all(color: Colors.yellow, width: 2) : null,
              ),
              child: const Text('Button', style: TextStyle(color: Colors.white, fontSize: 16)),
            );
          },
        ),
      ),
    );
  }
}
```

<details><summary>Button</summary>

```dart
import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:flutter/semantics.dart';

class _ButtonStateScope extends InheritedNotifier<WidgetStatesController> {
  const _ButtonStateScope({required super.notifier, required super.child});
}

extension type ButtonState._(Set<WidgetState> value) {
  bool get hovered => value.contains(WidgetState.hovered);
  bool get pressed => value.contains(WidgetState.pressed);
  bool get focused => value.contains(WidgetState.focused);
  bool get disabled => value.contains(WidgetState.disabled);
  Set<WidgetState> get states => value;
}

class Button extends StatefulWidget {
  const Button({
    super.key,
    this.onHover,
    this.onPressed,
    this.onFocusChange,
    this.focusNode,
    this.autofocus = false,
    this.behavior = HitTestBehavior.deferToChild,
    this.state,
    this.mouseCursor,
    required this.child,
  });

  final ValueChanged<bool>? onHover;
  final VoidCallback? onPressed;
  final ValueChanged<bool>? onFocusChange;
  final FocusNode? focusNode;
  final bool autofocus;
  final HitTestBehavior behavior;
  final Set<WidgetState>? state;
  final WidgetStateProperty<MouseCursor>? mouseCursor;
  final Widget child;

  static ButtonState of(BuildContext context) {
    final Set<WidgetState> value = context
        .dependOnInheritedWidgetOfExactType<_ButtonStateScope>()
        !.notifier
        !.value;
    return ButtonState._(value);
  }

  @override
  State<Button> createState() => _ButtonState();
}

class _ButtonState extends State<Button> {
  late final Map<Type, Action<Intent>> _actions = <Type, Action<Intent>>{
    ActivateIntent: CallbackAction<ActivateIntent>(onInvoke: _handleActivation),
    ButtonActivateIntent: CallbackAction<ButtonActivateIntent>(onInvoke: _handleActivation),
  };
  Map<Type, GestureRecognizerFactory>? _gestures;
  DeviceGestureSettings? _gestureSettings;

  // If a focus node isn't given to the widget, then we have to manage our own.
  FocusNode? _internalFocusNode;
  FocusNode get _focusNode => widget.focusNode ?? _internalFocusNode!;
  final WidgetStatesController _statesController = WidgetStatesController();

  bool get isHovered => _isHovered;
  bool _isHovered = false;
  set isHovered(bool value) {
    if (_isHovered != value) {
      _isHovered = value;
      _statesController.update(WidgetState.hovered, value);
    }
  }

  bool get isPressed => _isPressed;
  bool _isPressed = false;
  set isPressed(bool value) {
    if (_isPressed != value) {
      _isPressed = value;
      _statesController.update(WidgetState.pressed, value);
    }
  }

  bool get isFocused => _isFocused;
  bool _isFocused = false;
  set isFocused(bool value) {
    if (_isFocused != value) {
      _isFocused = value;
      _statesController.update(WidgetState.focused, value);
    }
  }

  bool get isEnabled => _isEnabled;
  bool _isEnabled = false;
  set isEnabled(bool value) {
    if (_isEnabled != value) {
      _isEnabled = value;
      _statesController.update(WidgetState.disabled, !value);
    }
  }

  @override
  void initState() {
    super.initState();
    if (widget.focusNode == null) {
      _internalFocusNode = FocusNode();
    }

    _statesController.value = widget.state ?? <WidgetState>{};
    isEnabled = widget.onPressed != null;
    isFocused = _focusNode.hasPrimaryFocus;
  }

  @override
  void didUpdateWidget(Button oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (widget.focusNode != oldWidget.focusNode) {
      if (widget.focusNode != null) {
        _internalFocusNode?.dispose();
        _internalFocusNode = null;
      } else {
        assert(_internalFocusNode == null);
        _internalFocusNode = FocusNode();
      }

      isFocused = _focusNode.hasPrimaryFocus;
    }

    if (widget.state != oldWidget.state && widget.state != null) {
      _statesController.value = widget.state!;
    }

    if (widget.onPressed != oldWidget.onPressed) {
      if (widget.onPressed == null) {
        isEnabled = isHovered = isPressed = isFocused = false;
      } else {
        isEnabled = true;
      }
    }
  }

  @override
  void dispose() {
    _statesController.dispose();
    _internalFocusNode?.dispose();
    _internalFocusNode = null;
    super.dispose();
  }

  void _handleFocusChange([bool? focused]) {
    isFocused = _focusNode.hasPrimaryFocus;
    widget.onFocusChange?.call(isFocused);
  }

  void _handleActivation([Intent? intent]) {
    isPressed = false;
    widget.onPressed?.call();
  }

  void _handleTapDown(TapDownDetails details) {
    isPressed = true;
  }

  void _handleTapUp(TapUpDetails? details) {
    isPressed = false;
    widget.onPressed?.call();
  }

  void _handleTapCancel() {
    isPressed = false;
  }

  void _handlePointerExit(PointerExitEvent event) {
    if (isHovered) {
      isHovered = false;
      widget.onHover?.call(false);
    }
  }

  void _handlePointerEnter(PointerEnterEvent event) {
    if (!isHovered) {
      isHovered = true;
      widget.onHover?.call(true);
    }
  }

  void _handleDismiss() {
    Actions.invoke(context, const DismissIntent());
  }

  @override
  Widget build(BuildContext context) {
    final DeviceGestureSettings? newGestureSettings = MediaQuery.maybeGestureSettingsOf(context);
    if (_gestureSettings != newGestureSettings) {
      _gestureSettings = newGestureSettings;
      _gestures = null;
    }

    _gestures ??= <Type, GestureRecognizerFactory>{
      TapGestureRecognizer: GestureRecognizerFactoryWithHandlers<TapGestureRecognizer>(
        () => TapGestureRecognizer(debugOwner: this),
        (TapGestureRecognizer instance) {
          instance
            ..onTapDown = _handleTapDown
            ..onTapUp = _handleTapUp
            ..onTapCancel = _handleTapCancel
            ..gestureSettings = _gestureSettings;
        },
      ),
    };

    final child = RawGestureDetector(
      behavior: widget.behavior,
      gestures: isEnabled ? _gestures! : const <Type, GestureRecognizerFactory>{},
      child: widget.child,
    );

    return MergeSemantics(
      child: Semantics.fromProperties(
        properties: SemanticsProperties(
          enabled: isEnabled,
          onDismiss: isEnabled ? _handleDismiss : null,
        ),
        child: Actions(
          actions: isEnabled ? _actions : <Type, Action<Intent>>{},
          child: Focus(
            autofocus: isEnabled && widget.autofocus,
            focusNode: _focusNode,
            canRequestFocus: isEnabled,
            skipTraversal: !isEnabled,
            onFocusChange: _handleFocusChange,
            child: _ButtonStateScope(
              notifier: _statesController,
              child: Builder(
                builder: (context) {
                  final MouseCursor? cursor = widget.mouseCursor?.resolve(
                    Button.of(context)?.states ?? <WidgetState>{},
                  );
                  return MouseRegion(
                    onEnter: isEnabled ? _handlePointerEnter : null,
                    onExit: isEnabled ? _handlePointerExit : null,
                    hitTestBehavior: HitTestBehavior.deferToChild,
                    cursor: cursor ?? MouseCursor.defer,
                    child: child,
                  );
                },
              ),
            ),
          ),
        ),
      ),
    );
  }
}

```
</details>

--- 

WidgetState is another good approach. I've also had luck animating WidgetState like so (let me know if you want to see the code):

```dart
Button(
      onPressed: () {
        onSectionTap(section);
      },
      decoration: WidgetStateDecoration.fromMap({
        WidgetState.pressed: AnimatedBoxDecoration(
          duration: const Duration(milliseconds: 100),
          color: Tail.gray200,
          border: depth < 3 ? border : null,
        ),
        WidgetState.any: AnimatedBoxDecoration(
          duration: const Duration(milliseconds: 100),
          color: Tail.lightTransparent,
          border: depth < 3 ? border : null,
        ),
      }),
//...
```

The only problem with this approach is that there is no good way to use simulations, and I use simulations a lot.

**Comment by TekExplorer** [comment]
...mayhaps the bot can be updated to understand what an umbrella issue is?

**Comment by mateusfccp-cos** [comment]
What about checkbox? Will we have a RawCheckbox or similar basic component decoupled from Material's `Checkbox`?
